benchmarks/discovery.md

```diff
@@ -0,0 +1,148 @@
+
+For the sake of reproducibility, we will re-run benchmarks with multiple router
+models to see if there is any signficant variation in results.
+
+## Software
+
+For mDNS, we will use the [Avahi](https://www.avahi.org/) open source mDNS
+software that is available as a Debian package.
+
+For SSDP, we will either:
+
+1. Fork and customize
+[libupnp](http://pupnp.sourceforge.net/) for the [Open Screen SSDP discovery
+protocol](../ssdp.md).
+2. Write our own SSDP implementation from scratch, possibly based on the
+   [Chromium](https://cs.chromium.org/chromium/src/chrome/browser/media/router/discovery/dial/dial_service.h)
+   implementation.
+
+As part of these implementations, we will instrument them to log data related to
+the Observables above, including (but not limited to):
+* Packets and bytes sent and received each second
+* Timestamp a device is discovered
+* Timestamp a device is detected to have left the network
+* Number of devices discovered
+
+A Python script will be used to start the sender and responder process with
+command line flags to adjust behavior for a given test condition. Data points of
+interest will be logged by the binaries to stdout, which will be injested and
+accumulated by the Python script.  At the end of the test, the script will write
+a JSON file to be copied to a Web server, so that a HTML frontend can be used to
+render nice charts of the results.  All of the scripts and source code for the
+benchmarks will be stored on GitHub.
+
+# Experimental Variables
+
+The following variables will be adjusted and the Observables collected for each
+condition.
+
+To keep the number of conditions within reason, one set of parameters will be
+adjusted and the others kept fixed.  In case the data shows interesting trends,
+we can run more conditions as desired as the entire process will be automated
+through scripting.
+
+## Protocols in use
+
+* SSDP alone
+* mDNS alone
+* SSDP and mDNS together
+
+## Device Population
+
+* Number of senders (devices initiating queries): 1, 2, 4, 8
+* Number of responders (devices responding to queries): 1, 2, 4, 8
+
+## Physical Positioning
+
+Most benchmarks will take place with the devices co-located (same room, same
+distance from router).
+
+However, to simulate signal fading effects, we will also collect data when the
+devices will be positioned in adjacent offices such that the senders are in one
+room separated by a wall from the router, and the responders are in a different
+room separated by a wall from the router.  This is to simulate a typical
+household where controller devices, receiver devices, and routers are spread
+among different rooms.
+
+**NOTE:** If there is interest in evaluating these protocols in particularly
+adverse WiFi conditions, we can investigate using a routing proxy to simulate
+additional congestion and packet loss, or simply move devices further away.
+
+## SSDP specific
+
+* Interval between `M-SEARCH` requests.
```

Comment:
Maybe we need also to consider the SSDP header `MX` which contains the maximum time in seconds (must be between 1 and 5 seconds) to delay `M-SEARCH` response. Also the `CACHE-CONTROL` header could be relevant.

Reply:
Good point - `MX` will definitely affect latency, added it as a variable.

`CACHE-CONTROL` does also impact the latency of discovery and the quantity of network traffic over a long time frame, since the value is suggested to be greater than 1800 seconds.  It could be interesting to vary it while making long term measurements over the course of hours (after discovery hits a steady state).  Added a note about steady state behavior as future work to the end.